File: apps/web/tests/recovery-http.test.mjs
import assert from 'node:assert/strict'
import {createServer} from 'node:http'
import {webcrypto} from 'node:crypto'
import test from 'node:test'
import {createTsLoader} from './helpers/load-ts.mjs'
test('recovery HTTP retains request key during refresh and exposes error data',async () => {
  const opens=[]
  let refreshes=0, query, snapshotQuery, readyBody
  const paths=[]
  const server=createServer(async(req,res)=>{
    const path=new URL(req.url,'http://local')
    paths.push(path.pathname)
    const reply=(data,status=200,reason)=>{res.writeHead(status,{'Content-Type':'application/json'});res.end(status===204?undefined:JSON.stringify({code:status,msg:reason,data}))}
    if(path.pathname.endsWith('/auth/refresh')) {refreshes++;reply({userId:7,token:'new'})}
    else if(path.pathname.endsWith('/sessions')) {
      let body='';for await(const chunk of req) body+=chunk
      opens.push({key:req.headers['idempotency-key'],body});reply({recoveryId:'session-a'},opens.length===1?401:200)
    } else if(path.pathname.endsWith('/mutations')) {
      query=Object.fromEntries(path.searchParams);reply({reason:'CURSOR_EXPIRED',floor:'9007199254740993',rebuildRequired:true},409,'CURSOR_EXPIRED')
    } else if(path.pathname.endsWith('/snapshot')) {snapshotQuery=Object.fromEntries(path.searchParams);reply({items:[]})}
    else if(path.pathname.endsWith('/cut')) reply({cut:'5'})
    else if(path.pathname.endsWith('/ready')) {let body='';for await(const chunk of req) body+=chunk;readyBody=JSON.parse(body);reply({ready:true})}
    else reply(null,204)
  })
  await new Promise(resolve=>server.listen(0,'127.0.0.1',resolve))
  const origin=`http://127.0.0.1:${server.address().port}`
  let session={userId:7,token:'old'}
  const {api}=createTsLoader({globals:{fetch,Headers,FormData,URLSearchParams,crypto:webcrypto,AbortController,setTimeout,clearTimeout,navigator:{userAgent:'MeshX test browser'}},
    replacements:{
      'apps/web/src/platform/nativeBridge.ts':{nativeBridge:{runtime:()=> 'web'}},
      'apps/web/src/platform/nativeTransport.ts':{nodeFetch:fetch},
      'apps/web/src/platform/nodeContext.ts':{currentNodeKey:()=>origin,currentNodeOrigin:()=>origin,currentNodeApiBasePath:()=>'/custom/api',apiUrl:path=>origin+'/custom/api'+path},
      'apps/web/src/utils/storage.ts':{readSession:()=>session,writeSession:value=>{session=value},clearSession:()=>{session=null}},
    }})('apps/web/src/services/api.ts')
  try {
    const input={protocolVersion:1,mode:'rebuild'}
    const opening=api.recovery.open(input,'same-logical-request');input.mode='mutated-after-call'
    assert.equal((await opening).recoveryId,'session-a')
    assert.equal(refreshes,1);assert.deepEqual(opens[0],opens[1]);assert.equal(opens.length,2)
    assert.equal(JSON.parse(opens[0].body).mode,'rebuild')
    await api.recovery.snapshot('session-a','opaque+token',200)
    assert.deepEqual(snapshotQuery,{pageToken:'opaque+token',limit:'200'})
    assert.equal((await api.recovery.cut('session-a')).cut,'5')
    await api.recovery.ready('session-a','9007199254740995',true)
    assert.deepEqual(readyBody,{appliedCursor:'9007199254740995',snapshotComplete:true})
    await assert.rejects(api.recovery.mutations('session-a','9007199254740992','9007199254740995',200),error=>error.code===409&&error.data.reason==='CURSOR_EXPIRED')
    assert.deepEqual(query,{after:'9007199254740992',through:'9007199254740995',limit:'200'})
    assert.equal(await api.recovery.release('session-a'),undefined)
    await assert.rejects(api.recovery.capabilities())
    assert.equal(paths.every(path=>path.startsWith('/custom/api/')),true)
  } finally {server.closeAllConnections();await new Promise(resolve=>server.close(resolve))}
})

File: apps/flutter-prototype/test/recovery_http_test.dart
import 'dart:convert';
import 'dart:io';
import 'package:flutter_test/flutter_test.dart';
import 'package:meshx_flutter_probe/data/meshx_api.dart';

void main() {
  test(
    'recovery HTTP preserves key through refresh, cursor precision and error data',
    () async {
      final server = await HttpServer.bind(InternetAddress.loopbackIPv4, 0);
      final origin = Uri.parse('http://127.0.0.1:${server.port}');
      final api = MeshXApi(origin);
      api.restoreCredentials({
        'origin': origin.toString(),
        'userId': 7,
        'nickname': 'probe',
        'token': 'old',
        'refreshCookie': 'test-refresh',
        'apiPath': '/custom/api',
        'wsPath': '/ws/chat',
        'nodeName': 'test',
      });
      final opens = <Map<String, Object?>>[];
      var refreshes = 0;
      Map<String, String>? query;
      final paths = <String>[];
      Map<String, String>? snapshotQuery;
      Object? readyBody;
      Future<void> reply(
        HttpRequest req,
        Object? data, {
        int status = 200,
        String? reason,
      }) async {
        req.response.statusCode = status;
        if (status != 204) {
          req.response.headers.contentType = ContentType.json;
          req.response.write(
            jsonEncode({'code': status, 'msg': reason, 'data': data}),
          );
        }
        await req.response.close();
      }

      server.listen((req) async {
        final path = req.uri.path;
        paths.add(path);
        if (path.endsWith('/auth/refresh')) {
          refreshes++;
          await reply(req, {'userId': 7, 'nickname': 'probe', 'token': 'new'});
        } else if (path.endsWith('/sessions')) {
          opens.add({
            'key': req.headers.value('Idempotency-Key'),
            'body': await utf8.decoder.bind(req).join(),
          });
          await reply(req, {
            'recoveryId': 'session-a',
          }, status: opens.length == 1 ? 401 : 200);
        } else if (path.endsWith('/mutations')) {
          query = req.uri.queryParameters;
          await reply(
            req,
            {
              'reason': 'CURSOR_EXPIRED',
              'floor': '9007199254740993',
              'rebuildRequired': true,
            },
            status: 409,
            reason: 'CURSOR_EXPIRED',
          );
        } else if (path.endsWith('/snapshot')) {
          snapshotQuery = req.uri.queryParameters;
          await reply(req, {'items': []});
        } else if (path.endsWith('/cut')) {
          await reply(req, {'cut': '5'});
        } else if (path.endsWith('/ready')) {
          readyBody = jsonDecode(await utf8.decoder.bind(req).join());
          await reply(req, {'ready': true});
        } else {
          await reply(req, null, status: 204);
        }
      });
      try {
        final input = <String, dynamic>{
          'protocolVersion': 1,
          'mode': 'rebuild',
        };
        final opening = api.openRecovery(
          input,
          idempotencyKey: 'same-logical-request',
        );
        input['mode'] = 'mutated-after-call';
        expect((await opening)['recoveryId'], 'session-a');
        expect(refreshes, 1);
        expect(opens.length, 2);
        expect(opens[0], opens[1]);
        expect(jsonDecode(opens[0]['body'] as String)['mode'], 'rebuild');
        await api.recoverySnapshot(
          'session-a',
          pageToken: 'opaque+token',
          limit: 200,
        );
        expect(snapshotQuery, {'pageToken': 'opaque+token', 'limit': '200'});
        expect((await api.recoveryCut('session-a'))['cut'], '5');
        await api.recoveryReady(
          'session-a',
          '9007199254740995',
          snapshotComplete: true,
        );
        expect(readyBody, {
          'appliedCursor': '9007199254740995',
          'snapshotComplete': true,
        });
        await expectLater(
          api.recoveryMutations(
            'session-a',
            '9007199254740992',
            '9007199254740995',
            limit: 200,
          ),
          throwsA(
            isA<ApiException>()
                .having((e) => e.code, 'status', 409)
                .having(
                  (e) => (e.data as Map)['reason'],
                  'reason',
                  'CURSOR_EXPIRED',
                ),
          ),
        );
        expect(query, {
          'after': '9007199254740992',
          'through': '9007199254740995',
          'limit': '200',
        });
        await api.releaseRecovery('session-a');
        await expectLater(
          api.recoveryCapabilities(),
          throwsA(isA<ApiException>()),
        );
        expect(paths.every((path) => path.startsWith('/custom/api/')), isTrue);
      } finally {
        api.close();
        await server.close(force: true);
      }
    },
  );
}
